src/components/selectable-table/SelectableTable.js

```diff
@@ -0,0 +1,155 @@
+/* eslint-disable jsx-a11y/no-noninteractive-element-to-interactive-role */
+/* eslint-disable jsx-a11y/click-events-have-key-events */
+import React, { useEffect, useCallback, useState } from 'react';
+import uuidv4 from 'uuid/v4';
+import union from 'lodash/union';
+import difference from 'lodash/difference';
+import isEqual from 'lodash/isEqual';
+
+/**
+ * A table with "selectable" rows.
+ * @prop {string[]} columns An array of column names, corresponding to data object properties.
+ * @prop {object[]} data An array of data objects used to populate table rows.
+ * @prop {function} onChange Callback function,
+ * passed a selection object when `allowMultiple` is false (and `null` if `allowUncheck` is true),
+ * or passed an array of selection objects when `allowMultiple` is true.
+ * @prop {string} idKey The key for a unique identifier property of `data` objects.
+ * @prop {string} valueKey If initially-selected rows are required,
+ * this key specifies a boolean property of the `data` objects
+ * indicating those rows that should be initially selected.
+ * @prop {boolean} allowMultiple Whether to allow multiple rows to be selected. By default, false.
+ * @prop {boolean} allowUncheck Whether to allow selected rows to be un-checked. By default, false.
+ * @prop {boolean} showTableHead Whether to show the table header element. By default, true.
+ */
+export default function SelectableTable(props) {
+  const {
+    columns,
+    data,
+    onChange,
+    idKey = 'id',
+    valueKey = 'value',
+    allowMultiple = false,
+    allowUncheck = false,
+    showTableHead = true,
+  } = props;
+
+  const [selectedRows, setSelectedRows] = useState(null);
+
+  // Callback function to update the `selectedRows` state.
+  const onSelectRow = useCallback((value, checked) => {
+    if (!allowMultiple) {
+      if (checked) {
+        setSelectedRows([value]);
+      } else if (!checked && allowUncheck) {
+        setSelectedRows([]);
+      }
+    } else if (checked) {
+      setSelectedRows(union(selectedRows || [], [value]));
+    } else if (!checked && allowUncheck) {
+      setSelectedRows(difference(selectedRows || [], [value]));
+    }
```

Comment:
```suggestion
    if (!allowMultiple) {
      if (checked) {
        setSelectedRows([value]);
      } else if (allowUncheck) {
        setSelectedRows([]);
      }
    } else {
      if (checked) {
        setSelectedRows(union(selectedRows || [], [value]));
      } else if (allowUncheck) {
        setSelectedRows(difference(selectedRows || [], [value]));
      }
    }
```

- In the `else` you know `!checked` will be true, so I don't think it's necessary.
- Perhaps wrap the second part in a single `else` to emphasize the parallels?
- I'm wondering about the case where neither `checked` nor `allowUncheck` is true... You could simplify the inner if-then-else, by adding a check outside:

```suggestion
  if (checked || allowUncheck) {
    if (!allowMultiple) {
      setSelectedRows(checked ? [value] : []);
    } else {
      setSelectedRows(
        checked
          ? union(selectedRows || [], [value])
          : difference(selectedRows || [], [value])
      );
    }
  }
``` 

I'm not sure this is an improvement...

Reply:
I like your second suggested change, and will adopt that!